content/development/testing.md

```diff
@@ -0,0 +1,147 @@
+---
+title: "Testing the Rule Set"
+#menuTitle: ""
+chapter: false
+weight: 8
+---
+
+# Testing for rule developers
+
+Well, you managed to write your rule, but now want to see if if can be added to the CRS? This document should help you to test it using the same tooling the project uses for its tests.
+
+Tests are performed using a Python tool called [ftw](https://github.com/coreruleset/ftw). We run them using a [GitHub actions pipeline](https://github.com/coreruleset/coreruleset/blob/v3.4/dev/.github/workflows/test.yml). You can easily reproduce that locally, in your workstation.
+
+For that you will need:
+
+- the coreruleset git repository
+- docker and docker-compose (modern versions of docker already include compose functionality)
+- python3
+- your rules and tests!
+
+## Setting up the basic environment
+
+I normally use [pipenv](https://docs.pipenv.org/) whenever Python is involved. It will give you both isolation and dependencies at once. But you can use basic pip and virtualenv and the result will be the same.
+
+For installing the python tooling, just use:
+
+```bash
+❯ pipenv install -r tests/regression/requirements.txt
```

Comment:
I would separate commands and output here, so that commands are easy to copy. Also remove prompt markers so they don't and up in the users' clipboard when copying. I'm not going to repeat this for every bash block here.

```suggestion
pipenv install -r tests/regression/requirements.txt
```

Reply:
I'm not confident I could do that without making a mistake.

Currently, the testing page is a copy-paste from the wiki article. I'm aware it is outdated and will need reworking anyway, so I'll make a note that the bash parts will need attention when the testing page is reworked.

I agree that it is not formatted well currently, but in the interests of getting this published I'm going to mark as 'resolved' for the moment.